.github/workflows/build.yml

```diff
@@ -0,0 +1,118 @@
+name: Build
+
+on: [push, pull_request]
+
+jobs:
+  generate-matrix:
+    runs-on: ubuntu-20.04
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.8'
+      - id: set-matrix
+        run: |
+          pip install platformio
+          python tools/ci/generate-matrix.py
+  build:
+    needs: generate-matrix
+    runs-on: ubuntu-20.04
+    strategy:
+      matrix: ${{ fromJson(needs.generate-matrix.outputs.matrix) }}
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.8'
+      - uses: actions/cache@v2
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-${{ hashFiles('requirements.txt') }}
+      - uses: actions/cache@v2
+        if: ${{ contains(matrix.env, 'esp32') }}
+        with:
+          path: ~/.platformio
+          key: ${{ runner.os }}-esp32-${{ hashFiles('platformio*.ini') }}
+      - uses: actions/cache@v2
+        if: ${{ contains(matrix.env, 'esp8266') }}
+        with:
+          path: ~/.platformio
+          key: ${{ runner.os }}-esp8266-${{ hashFiles('platformio*.ini') }}
+      - name: Dependencies
+        run: |
+          sudo apt install binutils
+          pip install -r requirements.txt
+          platformio update
+      - name: Build and archive
+        id: build-and-archive
+        env:
+          CHIP: ${{ matrix.chip }}
+          ENV: ${{ matrix.env }}
+        continue-on-error: true
```

Comment:
Also a repeated notice - this makes *every* build successful, job errors are treated as warnings in the overall view. 

There is https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast which will do part of what I wanted, but it seems build job generator still needs to be tweaked b/c the release job wants all jobs successful

Reply:
Yep I noticed.
Still we may need to check if it still will generate some artifacts on failed builds (not sure if we want it, but at least we need to know about it)

Reply:
Upload *could* happen in the same 'step', but right now it too depends on the success of the build step. Although, pygithub does not seem to know about Actions

My guess about the build status is that besides the context obj mentioned above, it should be available behind some kind of github API call. So, subsequent job can check how much of the matrix was completed (or were there any builds at all).

Also see https://github.com/platformio/platformio-core/issues/3858 related to the checkprogsize and staging esp8266 builds here